### ISSUE TITLE
Question: Performance of rds cache on lustreFS, SQLite cache on remote volume mounts.

### ISSUE THREAD

**Issue by mstr3336** [ropensci/drake#1025]
## Prework

- [x] Read and abide by `drake`'s [code of conduct](https://github.com/ropensci/drake/blob/master/CODE_OF_CONDUCT.md).
- [x] Search for duplicates among the [existing issues](https://github.com/ropensci/drake/issues), both open and closed.
- [x] Consider instead posting to [Stack Overflow](https://stackoverflow.com) under the [`drake-r-package` tag](https://stackoverflow.com/tags/drake-r-package).

## Question

Hi - I'm wondering what the performance of the rds cache on lustre file systems is, given the often stated limitation that [lustre doesn't like a lot of i/o with small files](https://www.nas.nasa.gov/hecc/support/kb/lustre-best-practices_226.html). 

From reading the `storrr` documentation - I did see that they specify that the small file performance of storrr may be optimized (to some extent) by performing write operations of the tiny content addressed .rds files in bulk/ in chunks

I don't know if drake's use of storr utilizes this optimization - nor if this optimization is appropiate to boost performance on a lustre file system (Lustre's bottleneck seems to stem from the way it partitions / does something with "stripes" for small files).

I wonder if anyone can comment towards this - I'd prefer to use the rds based cache, mainly because I'm able to mount the hpc filesystem locally, but I assume that reading from the database cache when mounting remotely would have to involve a full download of the cache to a temporary directory and a read from that.

Come to think of it- I'd also like someone to comment on this assumption, because I wouldn't mind continuing to use the SQL based cache if it's able to meet this requirement.

TL;DR

- [ ] Is the RDS based cache scalable/performant on lustre based file systems? 
  - [ ] Is the [bulk get/set ](http://richfitz.github.io/storr/articles/storr.html#bulk-getset) utilized by drake's backend? 
  - [ ] Does bulk get/set actually improve this (Maybe this is a better question to frame to lustre's maintainers/community) 
- [ ] Would reading from the SQLite database when mounted remotely (in my case SSHFS) incur significant overhead as a result of the remote hosting? 
  - After asking this, I now imagine that this is not a problem, since querying and pulling of remote data is one of the problems that relational database systems aims to solve.

**Comment by mstr3336** [comment]
If this would be more appropiate on the SO forum, I'm happy to post there, but since this is less of a question of usage and more a question of implementation mechanics I thought to post here

**Comment by wlandau** [comment]
> I'm wondering what the performance of the rds cache on lustre file systems is, given the often stated limitation that lustre doesn't like a lot of i/o with small files.

Great question. `drake`'s default `storr_rds()` cache does create a large collection of tiny files, but `drake::make()` tries not to detect, read, or write more than one or two files at a time. `drake_config()` and `outdated()`, however, could suffer.

My suspicion is that `drake` with the default `storr` will be slow on lustre and that there is not much we can do about it except maybe use `storr_dbi()`. You may be interested in the benchmarking workflow at https://github.com/wlandau/drake-examples/tree/master/overhead, downloadable with `drake_example("overhead")`. It requires profiling tools like `pprof` and `gperftools`, which can be hard to install, but you get super nice flame graphs. If we run it on lustre, we can start to investigate. I, unfortunately, do not have access to a lustre system.

@richfitz, do you know of anyone who has used heavily populated `storr`s on lustre systems?

> From reading the storr documentation - I did see that they specify that the small file performance of storr may be optimized (to some extent) by performing write operations of the tiny content addressed .rds files in bulk/ in chunks
>
> I don't know if drake's use of storr utilizes this optimization - nor if this optimization is appropiate to boost performance on a lustre file system (Lustre's bottleneck seems to stem from the way it partitions / does something with "stripes" for small files).

`drake` does not use `mget()` or `mset()` because of the memory concerns around large targets.

> I wonder if anyone can comment towards this - I'd prefer to use the rds based cache, mainly because I'm able to mount the hpc filesystem locally, but I assume that reading from the database cache when mounting remotely would have to involve a full download of the cache to a temporary directory and a read from that.

`storr_dbi()` + `RSQLite` allows you to maintain a local database. Just be sure to make sure `caching = "master" (default for `make()`) if you use parallel computing on a cluster.

> Is the RDS based cache scalable/performant on lustre based file systems?

Probably not, but let's find out with https://github.com/wlandau/drake-examples/tree/master/overhead. I, un

> Is the bulk get/set utilized by drake's backend?

No. `make()` operates on one target per process at a time, so `mget()` does not really make sense there. But maybe it could `drake_config()`

> Does bulk get/set actually improve this (Maybe this is a better question to frame to lustre's maintainers/community)

Possibly, though I cannot verify myself without access to a lustre system.

> Would reading from the SQLite database when mounted remotely (in my case SSHFS) incur significant overhead as a result of the remote hosting?

Yes, `make()` will slow down because of the high number of reads and writes over the network. But `storr_dbi()` + `RSQLite` can create *local* `storrs` usable by `drake`.

**Comment by mstr3336** [comment]
> My suspicion is that `drake` with the default `storr` will be slow on lustre and that there is not much we can do about it except maybe use `storr_dbi()`. You may be interested in the benchmarking workflow at https://github.com/wlandau/drake-examples/tree/master/overhead, downloadable with `drake_example("overhead")`. 

I'd be happy to follow this benchmarking workflow during my own investigations (Assuming I get round to it, I've had a fork of drake with a half-finished feature you told me I could implement 3 months ago 😢  ) - If you're interested in the result, I'd love to know if there's any particular output, format, or proof of work you'd like me to share with you (Maybe that's documented in the workflow already) .

> It requires profiling tools like `pprof` and `gperftools`, which can be hard to install, but you get super nice flame graphs. If we run it on lustre, we can start to investigate. I, unfortunately, do not have access to a lustre system.

My works' HPC runs on an old distro of centOS, and binaries are exposed by the sysadmins using `module load` etc, so I may have to use a singularity container to get anything requiring external binaries working, so we'll see. 

> `storr_dbi()` + `RSQLite` allows you to maintain a local database. Just be sure to make sure `caching = "master" (default for `make()`) if you use parallel computing on a cluster.
 
Fortunately, you made that very clear in your documentation, so I've already incorporated that into my HPC based workflow- This approach works very well for the HPC.

My local workflow tends to be nearly identical to my HPC workflow,  but I prefefer rds based cache because (I assume) there's loss bottleneck on a single worker thread trying to keep up with 23 other cores, and I don't have the same limitation on small file IO that lustre imposes. 

My intention isn't actually to run my workflow using a local machine on the network mounted - I'll clarify in my final response.

> Yes, `make()` will slow down because of the high number of reads and writes over the network. But `storr_dbi()` + `RSQLite` can create _local_ `storrs` usable by `drake`.

So my intention isn't actually to run the build locally against the remote cache, my local workflow uses a subset of the data & a local cache - I'm actually just interested in providing my team a user-friendly mechanism to perform interactive exploration / analysis of targets produced by the hpc using `drake::readd()` on the mounted cache. 

My project is a health infomatics project with rigid ethics around data security and storage, so data must be hosted on a secure remote disk for any long term purpose.

The performance tradeoff of network IO is acceptable to me, as long as that tradeoff is the overhead of reading/writing the *specific targets* I'm querying with `drake::readd()`.

So my question is then - if I want to read a target of size *n* from a cache of size _**N**_ is that cost scaled mostly linearly with *n* , or with _**N**_ ?

**Comment by wlandau** [comment]
>  If you're interested in the result, I'd love to know if there's any particular output, format, or proof of work you'd like me to share with you (Maybe that's documented in the workflow already) .

Thanks! The flame graphs from `pprof` are super helpful.

> My local workflow tends to be nearly identical to my HPC workflow, but I prefefer rds based cache because (I assume) there's loss bottleneck on a single worker thread trying to keep up with 23 other cores, and I don't have the same limitation on small file IO that lustre imposes.

If `caching = "master"` in `make()`, only the master process reads and writes targets, so even the DB cache should be thread safe (but `jobs_preprocess` needs to be 1). But if `caching = "worker"`, the workers store the targets, which means the RDS `storr` (or another thread safe `storr`) is required.

> So my intention isn't actually to run the build locally against the remote cache, my local workflow uses a subset of the data & a local cache - I'm actually just interested in providing my team a user-friendly mechanism to perform interactive exploration / analysis of targets produced by the hpc using drake::readd() on the mounted cache.

`readd()` only reads one target (one file) at a time, so that should be okay on lustre. `loadd()` with many targets, on the other hand, might be slow.

> The performance tradeoff of network IO is acceptable to me, as long as that tradeoff is the overhead of reading/writing the specific targets I'm querying with drake::readd().
>
> So my question is then - if I want to read a target of size n from a cache of size N is that cost scaled mostly linearly with n , or with N ?

I expect the complexity of `readd()` to be `O(n)`. It pretty much just calls `get()` with the `storr`.

https://github.com/ropensci/drake/blob/3af085d19c699da9b1d29681d6a4a541f7193d4f/R/cache.R#L74-L108

**Comment by wlandau** [comment]
Closing the issue, but not the thread. Let's keep the discussion going. I do not mean to shut down the conversation, I just think the closest thing we will have to a solution to this potential performance issue is #685, which will dramatically reduce the number of small metadata files if used correctly.

**Comment by wlandau** [comment]
Related: #1095.

**Comment by mstr3336** [comment]
Circling back to this issue - I'm curious about what `storr` / `DBI` backends people have used successfully with `drake`. 

`SQLite` has some scaling limitations, and I'm wondering if a switch to more heavyweight `DBI` compliant databases could be wise, and whether the spike in complexity of managing a non-embedded database in a workflow is viable / worth it.

I can't really find any exhaustive documentation on what people have tried, what's been successful, and what's been difficult about using cache drivers besides the ubiquitous `rds` and `RSQLite` backends. 

I'm considering `PostgresSQL`, and I also noticed `DuckDB`, both of which have `DBI` bindings - `DuckDB` has the advantage of being embeddable like `RSQLite`, but I've heard may have an easier time scaling to many columns. 

I've not seen anyone document experience with either of these backends - nor many other backends for that matter, and I'm wondering whether it's been done/ whether there's problems.
